File: epi/TreePreorder.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
public class TreePreorder {

  private static class NodeAndState {
    public BinaryTreeNode<Integer> node;
    public Boolean nodeProcessed;

    public NodeAndState(BinaryTreeNode<Integer> node, Boolean nodeProcessed) {
      this.node = node;
      this.nodeProcessed = nodeProcessed;
    }
  }

  @EpiTest(testDataFile = "tree_preorder.tsv")
  public static List<Integer> preorderTraversal(BinaryTreeNode<Integer> tree) {
    /*
      T O(N) S O(h) h is logN - N
     */
    List<Integer> rtList = new ArrayList<>();

    Deque<BinaryTreeNode<Integer>> stack = new ArrayDeque<>();
    BinaryTreeNode<Integer> root = tree;

    while(!stack.isEmpty() || root != null) {
      while(root != null) {
        rtList.add(root.data);
        stack.push(root);
        root = root.left;
      }

      BinaryTreeNode tmp = stack.pop();
      root = tmp.right;
    }

    return rtList;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "TreePreorder.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/TreePreorder.kt
package epi.kt

import epi.BinaryTreeNode
import java.util.ArrayDeque

object TreePreorder {
    //T O(N) S O(h)
    fun preorderTraversal(tree : BinaryTreeNode<Int>?) : List<Int> {
        var rtList = ArrayList<Int>()

        var stack = ArrayDeque<BinaryTreeNode<Int>>()
        var tmp = tree

        //left -> root -> right
        while(!stack.isEmpty() || tmp != null) {
            while(tmp != null) {
                rtList.add(tmp.data)
                stack.push(tmp)
                tmp = tmp.left
            }

            val node = stack.pop()
            tmp = node.right
        }

        return rtList
    }
}